crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -0,0 +1,1036 @@
+use std::{fmt, ops::Not, ptr::NonNull};
+
+#[derive(Debug)]
+pub enum LinkedChunkError {
+    InvalidChunkIndex { index: ChunkPosition },
+    ChunkIsAGap { index: ChunkPosition },
+    InvalidItemIndex { index: usize },
+}
+
+pub struct LinkedChunk<T, const CHUNK_CAPACITY: usize> {
+    first: NonNull<Chunk<T, CHUNK_CAPACITY>>,
+    last: Option<NonNull<Chunk<T, CHUNK_CAPACITY>>>,
+    length: usize,
+}
+
+impl<T, const C: usize> LinkedChunk<T, C> {
+    /// Create a new [`Self`].
+    pub fn new() -> Self {
+        Self { first: Chunk::new_items_leaked(), last: None, length: 0 }
+    }
+
+    /// Get the number of items in this linked chunk.
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    /// Push items at the end of the [`LinkedChunk`], i.e. on the last
+    /// chunk.
+    ///
+    /// If the last chunk doesn't have enough space to welcome all `items`,
+    /// then new chunks can be created (and linked appropriately).
+    pub fn push_items_back<I>(&mut self, items: I)
+    where
+        I: IntoIterator<Item = T>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        let items = items.into_iter();
+        let number_of_items = items.len();
+
+        let last_chunk = unsafe { self.last.as_mut().unwrap_or(&mut self.first).as_mut() };
+
+        // Push the items.
+        let last_chunk = last_chunk.push_items(items);
+
+        assert!(last_chunk.is_last_chunk(), "`last_chunk` must be… the last chunk");
```

Comment:
I don't understand: do you want multiple `assert_…` method, or just a single one that checks every possible situations?

Reply:
I thought of a single method that would check post-invariants and that we could call at the end of the functions editing the `LinkedChunk`'s state. (Not sure if it's a useful idea, you tell me — I thought it would be super nice, would we start fuzzing this :eyes: )

Reply:
The tests are already checking many invariants, so I suppose it's fine. There isn't many neither.